src/index.ts

```diff
@@ -58,8 +66,74 @@ const formatDnsDidRecord = ({ a, v, p, type }: { [key: string]: string }) => {
   };
 };
 
-export const queryDns = async (domain: string): Promise<IDNSQueryResponse> => {
-  const { data } = await axios.get<IDNSQueryResponse>(`https://dns.google/resolve?name=${domain}&type=TXT`);
+export const queryGoogleDns = async (domain: string, headers: Record<string, string>): Promise<IDNSQueryResponse> => {
+  const { data } = await axios.get<IDNSQueryResponse>(`https://dns.google/resolve?name=${domain}&type=TXT`, {
+    headers,
+  });
+
+  return data;
+};
+
+export const queryCloudflareDns = async (
+  domain: string,
+  headers: Record<string, string>
+): Promise<IDNSQueryResponse> => {
+  const { data } = await axios.get<IDNSQueryResponse>(`https://1.1.1.1/dns-query?name=${domain}&type=TXT`, { headers });
+  return data;
+};
+
+export const parseCustomDns = (customDns: CustomDNS): AxiosRequestConfig => {
+  const fullPath = `${customDns.url}${
+    customDns.params
+      ? `?${Object.entries(customDns.params)
+          .reduce((acc, [k, v]) => {
+            acc += `${k}=${v}&`;
+            return acc;
+          }, "")
+          .slice(0, -1)}`
+      : ""
+  }`;
+  const { body, headers } = customDns;
+  const method: Method = customDns.method || "GET";
+
+  const config = {
+    method,
+    url: fullPath,
+  };
+
+  if (body) {
+    Object.assign(config, { data: body });
+  }
+
+  if (headers) {
+    Object.assign(config, { headers });
+  }
+
+  return config;
+};
+
+export const queryCustomDns = async (customDns: CustomDNS): Promise<IDNSQueryResponse> => {
+  const config = parseCustomDns(customDns);
+
+  const { data } = await axios(config);
+
+  return data;
+};
+
+export const queryDns = async (domain: string, customDns?: CustomDNS): Promise<IDNSQueryResponse> => {
```

Comment:
this is going in the right direction IMHO, but I think we should go one step further:
1. allow customDns to be an array
2. Make our code use customDns option, instead of hardcoding the dependency to google and cloudflare

I would add another suggestion:
Maybe we could pass resolver instead: A resolver is a function that takes a domain and return a <IDNSQueryResponse>
A resolver is better in the way it gives more control to the implementer regarding what the API returns. I'm not sure whether DoH always return the same response (same shape) across the different provider, but a resolver helps to fix that issue.
To help people build their own resolver and if data shape is always the same, we can export the function that parses the result and transform into a IDNSQueryResponse